src/geodesy.jl

```diff
@@ -11,43 +11,62 @@ import Base.isless
 import Base.isgreater
 import DataStructures.Queue, DataStructures.Deque, DataStructures.enqueue!, DataStructures.dequeue!, DataStructures.dequeue_pair!
 
+_build_connectivity(r::Int, n) = trues(ntuple(_->r, n))
+function _build_connectivity(r::NTuple{N,Int}, n) where N
+    N <= n || throw(ArgumentError("radius length $(length(r)) must be less or equal to $n"))
+    sz = (r..., ntuple(_->1, n-N)...)
+    return trues(sz)
+end
 
 """
-    hmaxima(image, connectivity=[], constant) -> Array
-The h-maxima transformation suppresses all maxima whose "relative" depth is below a given threshold (constant)
-# Arguments
-- `image::AbstractArray{T, N}`: where {T<:ImageCore.NumberLike , N} the ``N``-dimensional input image
-- `connectivity::AbstractArray{Bool}`: the neighborhood connectivity.
-- `constant::T`: the soustracted constant.
-# Returns
-An array of the same type and shape as the `image`.
-# See also
-[`hminima`](@ref)
-# References
-    Morphological image analysis by Soille pg 170-172
+    hmaximum(img, h; r=3)
+
+For grayscale image `img`, the h-maximum transformation suppresses all regional maxima whose
+depth is below a given threshold level `h`. The regional size `r` can be either integer or
+tuple of integer.
+
+```julia
+using ImageMorphology, TestImages
+
+img = Gray.(testimage("blob"))
+img_h = hmaximum(img, 0.4)
+
+# suppress the regional maxima along the second dimension
+img_h = hmaximum(img, 0.4; r=(1, 3))
+```
+
+See also [`hminimum`](@ref).
 """
-function hmaxima(image::AbstractArray{T,N}, connectivity::AbstractArray{Bool}, constant::T) where {T<:ImageCore.NumberLike,N}
-    tmp = image .- constant
-    return underbuild(tmp, image, connectivity)
+function hmaximum(img::AbstractArray, h; r=3)
+    # TODO(johnnychen94): support in-place version
+    connectivity = _build_connectivity(r, ndims(img))
+    tmp = mappedarray(x -> x - h, img)
```

Comment:
we save a temporary ?

Reply:
Yes, `mappedarray` from MappedArrays builds a lazy array container so we save a memory allocation.